packages/http-server/src/__tests__/body-params-validation.spec.ts

```diff
@@ -0,0 +1,277 @@
+import getHttpOperations from '@stoplight/prism-cli/src/util/getHttpOperations';
+import { createLogger } from '@stoplight/prism-core';
+import fastify = require('fastify');
+import { relative, resolve } from 'path';
+import { createServer } from '../';
+import { IPrismHttpServer } from '../types';
+
+const logger = createLogger('TEST', { enabled: false });
+
+async function instantiatePrism(fixture: string) {
+  const path = relative(process.cwd(), resolve(__dirname, 'fixtures', fixture));
+  const operations = await getHttpOperations(path);
+  const server = createServer(operations, { components: { logger }, config: { mock: { dynamic: false } } });
+  return server;
+}
+
+describe('body params validation', () => {
+  let server: IPrismHttpServer;
+
+  afterAll(async () => {
+    await server.fastify.close();
+  });
+
+  describe('oas3 with encodings', () => {
+    beforeEach(async () => {
+      server = await instantiatePrism(`encodings.oas3.yaml`);
+    });
+
+    // Ref: https://github.com/stoplightio/prism/issues/496
+    test.skip('allowReserved set to true', async () => {
+      const response = await server.fastify.inject({
+        method: 'POST',
+        url: '/allowReserved',
+        payload: {
+          reserved: ":/?#[]@!$&'()*+,;",
+        },
+      });
+
+      expect(response.statusCode).toEqual(422);
+    });
+  });
+
+  describe.each([['oas2'], ['oas3']])('%s with body param', oas => {
```

Comment:
Integration tests are for making sure values are being passed around correctly and the boundaries between code work properly. E2E is more overview “does the whole thing work”, but struggles with intricate assertions or multiple similar-but-slightly-tweaked scenarios like this.

Reply:
Aaaactually yeah this is a bit tough to see what the test is doing because it’s all coming from a file. I see what you mean about feeling like an E2E test @XVincentX, this might be a bit too zoomed out.

Reply:
I did this for one particular reason to be honest: debugging :) It's a PAIN to debug test harness but with these tests I was able to spot the issue in a matter of minutes. 

Yup it's all coming from a file because of how the loader works (loads stuff from files :) )
It would be great to have this stuff inline I agree, maybe we'll refactor at some point later when we support 'passing pure specs'.

E2E means testing, like Phil said, is testing if the **whole thing works** from a user perspective. Since Prism is both an `exe` file and and API e2e is a thin line. I would like to keep those tests in though because they test more than what I ended up with in test-harness.

Reply:
@chris-miaskowski I am not advocating for removing these tests, I am advocating for making them more readable by doing less. Instead of using the loader surely we can just pass in operations, right? that's the benefit of not requiring graphite, we dont have to load a file we can pass in operations. 

Then this becomes more of.an integration test, because we are checking to see how operations can effect the various classes involved which result in a HTTP response being generated. And the readability of each single test will be greatly improved because you dont have to open up another file to read what its trying to do. win win win.

Reply:
@philsturgeon ok gotcha. Sounds like a good idea, **but...**! The problem with using pure http-operations is that we'll not spot issues related to transforming oas2/3 to http-operations.

These tests were intended to test the actual specs (vs the abstracted http-operation). I agree that doing the latter will match more to the definition of integration test but at the same time will cover less stuff.

Reply:
@chris-miaskowski I understand. This might sound like the concern is about purity of tests to a textbook definition, but its more that these conventions exist for a reason, a reason that we would benefit from.  "Covering more stuff" is one way to look at it, but another way would be "less targeted and therefore less clear about what we're actually trying to test."

E2E is for catching overall problems, integration tests should be trying to put code into the appropriate state to trigger the reaction expected. More tests can be written to target the other specific things, but trying to do it all at once just means we are making confusing tests that are spooky to touch because what are we asserting. 

Accidental coverage should not be our goal here.

Reply:
@philsturgeon let's not forget about debuggability. We can ditch the files, convert all specs to http-operations and call it a day. We loose the subtle differences between oas2 and oas3 this way which we're hoping to get covered by e2e. In reality though it's much slower to cover them with e2e than this "semi-integration semi whatever" test. 

I'm not going to try to defend this though :) If you folks think it makes more sense to loose some coverage and ease of development in favor of readability then let it be.

@XVincentX @philsturgeon if you both give thumbs up I will convert :)

Reply:
@chris-miaskowski unit tests on transformers, integration tests that are aimed at that specific goal, and the end to end tests cover this. Again, accidental coverage should not be our goal here.

Reply:
I stand with @philsturgeon here — if the observability/debug experience is the main issue of writing them here, then we might need to address this somehow.

One of the things that I think would be beneficial would be a flag where you can either run the Harness on the final executable **OR** on the "source code" directly. This should improve the debuggbility significantly. Does that address your concern?

Reply:
@XVincentX @philsturgeon updated that test. Looks good now?

Reply:
> One of the things that I think would be beneficial would be a flag where you can either run the Harness on the final executable OR on the "source code" directly. This should improve the debuggbility significantly. Does that address your concern?

I don't think we should do that. E2E should always run on binary. We have `jest` to run on source code.